lib/puppet/functions/step.rb

```diff
@@ -0,0 +1,88 @@
+# Provides stepping with given interval over elements in an iterable and optionally runs a
+# [lambda](http://docs.puppetlabs.com/puppet/latest/reference/lang_lambdas.html) for each
+# element.
+#
+# This function takes two to three arguments.
+#
+# 1. An iterable that the function will iterate over.
+# 2. An integer step factor. This must be a positive integer.
+# 3. An optional lambda, which the function calls for each element in the interval. It must
+# request one parameter.
+#
+# @example Using the `step` function
+#
+# `$data.step(<n>) |$parameter| { <PUPPET CODE BLOCK> }`
+#
+# or
+#
+# `$stepped_data = $data.step(<n>)`
+#
+# or
+#
+# `step($data, <n>) |$parameter| { <PUPPET CODE BLOCK> }`
+#
+# or
+#
+# `$stepped_data = step($data, <n>)`
+#
+# When no third argument is present, Puppet returns an iterable that represents the nth element
+# of its first argument. This allows methods on iterables to be chained.
+#
+# When a lamdba is given as the third argument, Puppet iterates the first argument and passes each
+# nth value in turn to the lambda, then returns the first argument unchanged.
+#
+# @example Using the `step` function with an array, a step factor, and a one-parameter lambda
```

Comment:
here lambda is fine, since it is on the calling side, the mechanism with which a block is given to the function. (Pointing this out just to avoid the otherwise obvious question).